Source/Scene/ModelExperimental/PickingPipelineStage.js

```diff
@@ -0,0 +1,123 @@
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+
+/**
+ * The picking pipeline stage is responsible for handling picking of primitives.
+ *
+ * @namespace PickingPipelineStage
+ * @private
+ */
+var PickingPipelineStage = {};
+
+/**
+ * Process a primitive. This modifies the following parts of the render resources:
+ * <ul>
+ *  <li>adds attribute and varying declaration for the pick color vertex attribute in the vertex shader for instanced meshes</li>
+ *  <li>adds declaration for the pick color uniform for non-instanced meshes</li>
+ *  <li>adds defines in the shader for when picking is enabled</li>
+ *  <li>creates the pick ID objects in the context</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} node The primitive.
+ * @param {FrameState} frameState The frame state.
+ */
+PickingPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  var context = frameState.context;
+  var runtimeNode = renderResources.runtimeNode;
+  var shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine("ALLOWS_PICKING");
+
+  if (defined(runtimeNode.node.instances)) {
+    // For instanced meshes, a pick color vertex attribute is used.
+    processInstancedPickIds(renderResources, context);
+  } else {
+    // For non-instanced meshes, a pick color uniform is used.
+    var model = renderResources.model;
+    var pickObject = {
+      model: model,
+      node: renderResources.runtimeNode,
+      primitive: renderResources.runtimePrimitive,
+    };
+
+    var pickId = context.createPickId(pickObject);
+    model._resources.push(pickId);
+    shaderBuilder.addUniform(
+      "vec4",
+      "czm_pickColor",
+      ShaderDestination.FRAGMENT
+    );
+
+    var uniformMap = renderResources.uniformMap;
+    uniformMap.czm_pickColor = function () {
+      return pickId.color;
+    };
+
+    renderResources.pickId = "czm_pickColor";
+  }
+};
+
+function processInstancedPickIds(renderResources, context) {
+  var instanceCount = renderResources.instanceCount;
+  var pickIds = new Array(instanceCount);
+  var pickIdsTypedArray = new Uint8Array(instanceCount * 4);
+
+  var modelResources = renderResources.model._resources;
+  for (var i = 0; i < instanceCount; i++) {
+    var pickObject = {
+      model: renderResources.model,
+      node: renderResources.runtimeNode,
+      primitive: renderResources.runtimePrimitive,
+      instance: {
+        id: i,
+      },
+    };
+
+    var pickId = context.createPickId(pickObject);
+    modelResources.push(pickId);
+    pickIds[i] = pickId;
+
+    var pickColor = pickId.color;
+    pickIdsTypedArray[i * 4 + 0] = Color.floatToByte(pickColor.red);
+    pickIdsTypedArray[i * 4 + 1] = Color.floatToByte(pickColor.green);
+    pickIdsTypedArray[i * 4 + 2] = Color.floatToByte(pickColor.blue);
+    pickIdsTypedArray[i * 4 + 3] = Color.floatToByte(pickColor.alpha);
```

Comment:
Is this pattern used anywhere else? just curious if adding a `Color.packBytes()` method would be helpful. I see that `pack()` won't work here since you want actual values from 0-255.

Reply:
See https://github.com/CesiumGS/cesium/blob/404c4e8e478f1fb306dc1187845ab41ed97564d5/Source/Scene/ModelInstanceCollection.js#L629-L632

Reply:
I also see this pattern in several files including:

- ColorGeometryInstanceAttribute
- GeometryAttribute
- GeometryInstanceAttribute
- PolylineGeometry
- SimplePolylineGeometry (a few times)

I think a `Color.packBytes()` would be nice here. Though given the scope of changes, it would be best to do this as a separate PR. Or, if that's too much of a hassle for now, at least open an issue.